### PR TITLE
[MB-1327] Remove read permission for jks when ssl is enabled

### PR DESCRIPTION
**Pull request by sajinidesilva**
jira: https://wso2.org/jira/browse/MB-1327

### CODE REVIEW

**Comment by sajinidesilva** [comment]
Related jira https://wso2.org/jira/browse/MB-1327

File: modules/andes-core/client/src/main/java/org/wso2/andes/client/AMQConnection.java [hunk 1]
```diff
@@ -610,16 +622,49 @@ public boolean attemptReconnection()
...
+                }
+
+            });
+            return returnObject;
+        } catch (PrivilegedActionException e) {
+            if (e.getException() instanceof IOException) {
+                throw (IOException) e.getException();
+            } else if (e.getException() instanceof AMQException) {
+                throw (AMQException) e.getException();
+            }
```
Comment (ramith) [inline modules/andes-core/client/src/main/java/org/wso2/andes/client/AMQConnection.java]: shouldn't we be propagating this to higher level? ( than returning null)?

File: modules/andes-core/client/src/main/java/org/wso2/andes/client/AMQConnection.java [hunk 2]
```diff
@@ -610,16 +622,49 @@ public boolean attemptReconnection()
...
 
-    public <T, E extends Exception> T executeRetrySupport(FailoverProtectedOperation<T,E> operation) throws E
-    {
-        return _delegate.executeRetrySupport(operation);
+    public <T, E extends Exception> T executeRetrySupport(final FailoverProtectedOperation<T, E> operation) throws E {
+        // Requires permission java.lang.RuntimePermission "modifyThreadGroup"
+        Object returnObject = null;
+        try {
+            returnObject = AccessController.doPrivileged(new PrivilegedExceptionAction<Object>() {
+                public Object run() throws E {
```
Comment (ramith) [inline modules/andes-core/client/src/main/java/org/wso2/andes/client/AMQConnection.java]: wrong use of generics.

Comment (ramith) [inline modules/andes-core/client/src/main/java/org/wso2/andes/client/AMQConnection.java]: if something goes wrong we are returning a null... is that correct?

File: modules/andes-core/common/src/main/java/org/wso2/andes/transport/ConnectionSettings.java [hunk 3]
```diff
@@ -155,6 +156,15 @@ public void setUseSSL(boolean useSSL)
         this.useSSL = useSSL;
     }
 
+    public void loadSSLConfigFromSysConfig(){
```
Comment (ramith) [inline modules/andes-core/common/src/main/java/org/wso2/andes/transport/ConnectionSettings.java]: a method comment would be nice ( explaining the order we load these values etc)

File: modules/andes-core/client/src/main/java/org/wso2/andes/client/AMQConnection.java [hunk 4]
```diff
@@ -610,16 +622,49 @@ public boolean attemptReconnection()
...
+
+                    return _delegate.executeRetrySupport(operation);
+
+                }
+            });
+        } catch (PrivilegedActionException e) {
+            throw (E) e.getException();
+        } finally {
+            return (T) returnObject;
```
Comment (ramith) [inline modules/andes-core/client/src/main/java/org/wso2/andes/client/AMQConnection.java]: shall we re-arrange how we return 'returnObject' ?

**Comment by ramith** [comment]
git rebase --interfactive HEAD~3 would be a nice thing to do.

File: modules/andes-core/client/src/main/java/org/wso2/andes/client/AMQConnection.java [hunk 5]
```diff
@@ -610,16 +622,48 @@ public boolean attemptReconnection()
...
     }
 
-    public ProtocolVersion makeBrokerConnection(BrokerDetails brokerDetail) throws IOException, AMQException
+    public ProtocolVersion makeBrokerConnection(final BrokerDetails brokerDetail) throws IOException, AMQException
     {
-        return _delegate.makeBrokerConnection(brokerDetail);
+        // Requires permission java.lang.RuntimePermission "modifyThreadGroup"
+        try {
+            ProtocolVersion returnObject = AccessController.doPrivileged(new PrivilegedExceptionAction<ProtocolVersion>() {
+                public ProtocolVersion run() throws IOException, AMQException{
```
Comment (ramith) [inline modules/andes-core/client/src/main/java/org/wso2/andes/client/AMQConnection.java]: wrong indentation.

File: modules/andes-core/client/src/main/java/org/wso2/andes/client/AMQConnection.java [hunk 6]
```diff
@@ -880,8 +924,14 @@ private void doClose(List<AMQSession> sessions, long timeout) throws JMSExceptio
 
 	                    closeAllSessions(null, timeout, startCloseTime);
 
-                        //This MUST occur after we have successfully closed all Channels/Sessions
-                        _taskPool.shutdown();
+                        // Requires permission java.lang.RuntimePermission "modifyThread"
+                        AccessController.doPrivileged(new PrivilegedAction<Void>( ){
```
Comment (ramith) [inline modules/andes-core/client/src/main/java/org/wso2/andes/client/AMQConnection.java]: wrong indentation.

File: modules/andes-core/client/src/main/java/org/wso2/andes/client/AMQConnectionDelegate_8_0.java [hunk 7]
```diff
@@ -116,6 +116,14 @@ public ProtocolVersion makeBrokerConnection(BrokerDetails brokerDetail) throws A
             settings.setTrustStoreCertType(sslConfig.getCertType());
         }
 
+        if(settings.getKeyStorePath() == null && settings.getKeyStorePassword() ==null && settings
```
Comment (ramith) [inline modules/andes-core/client/src/main/java/org/wso2/andes/client/AMQConnectionDelegate_8_0.java]: seems we are missing a space here.

File: modules/andes-core/common/src/main/java/org/wso2/andes/transport/network/security/ssl/SSLUtil.java [hunk 8]
```diff
@@ -163,19 +163,19 @@ public static KeyStore getInitializedKeyStore(String storePath, String storePass
...
-            if (in == null)
-            {
-                throw new IOException("Unable to load keystore resource: " + storePath);
+
+            in = Thread.currentThread().getContextClassLoader().getResourceAsStream(storePath);
+
+            if (null == in) { // Resource was not found in the classpath. Try to load it from the file system.
+                File f = new File(storePath);
+                if (f.exists() && f.canRead())
+                {
```
Comment (a5anka) [inline modules/andes-core/common/src/main/java/org/wso2/andes/transport/network/security/ssl/SSLUtil.java]: Shall we move the curly brace to previous line to correct code formatting?

**Pull request closed without merging**